mailpoet/assets/js/src/automation/listing/index.tsx

```diff
@@ -1,17 +1,90 @@
 import { Search, TableCard } from '@woocommerce/components/build';
+import { TabPanel } from '@wordpress/components';
 import { __ } from '@wordpress/i18n';
+import { useHistory, useLocation } from 'react-router-dom';
 import { getRow } from './get-row';
-import { Workflow } from './workflow';
+import { Workflow, WorkflowStatus } from './workflow';
 
 type Props = {
   workflows: Workflow[];
   loading: boolean;
 };
 
+type FilterTab = {
+  name: string;
+  title?: JSX.Element;
+  classname: string;
+};
+
 export function AutomationListing({ workflows, loading }: Props): JSX.Element {
-  const rows = workflows.map((workflow) => getRow(workflow));
+  const history = useHistory();
+  const location = useLocation();
+  const pageSearch = new URLSearchParams(location.search);
+
+  const updateUrlSearchString = (keysAndValues: Record<string, string>) => {
+    Object.keys(keysAndValues).forEach((key) => {
+      pageSearch.set(key, keysAndValues[key]);
+      if (['per_page', 'status'].includes(key)) {
+        pageSearch.set('paged', '1');
+      }
+    });
+    history.replace({ search: pageSearch.toString() });
+  };
+
+  const groupedWorkflows = {};
+  workflows.forEach((workflow) => {
+    if (!groupedWorkflows[workflow.status]) {
+      groupedWorkflows[workflow.status] = [];
+    }
+    groupedWorkflows[workflow.status].push(workflow);
+  });
+
+  const tabLabels = [
+    { status: WorkflowStatus.ACTIVE, label: 'Active' },
+    { status: WorkflowStatus.INACTIVE, label: 'Inactive' },
+    { status: WorkflowStatus.DRAFT, label: 'Draft' },
+    { status: WorkflowStatus.TRASH, label: 'Trash' },
+  ];
 
-  const headers = [
+  const tabs: FilterTab[] = [
+    {
+      name: 'all',
+      title: (
+        <>
+          <span>All</span>{' '}
+          <span className="count">{workflows.length.toString()}</span>
+        </>
+      ),
+      classname: 'tab-all',
+    },
+  ];
+
+  tabLabels.forEach((tabInfo) => {
+    const tab: FilterTab = {
+      name: tabInfo.status,
+      classname: `tab-${tabInfo.status}`,
+    };
+
+    if (
+      groupedWorkflows[tabInfo.status] &&
+      groupedWorkflows[tabInfo.status].length > 0
+    ) {
+      tab.title = (
+        <>
+          <span>{tabInfo.label}</span>{' '}
+          <span className="count">
+            {groupedWorkflows[tabInfo.status].length}
+          </span>
+        </>
+      );
+    } else {
+      tab.title = <span>{tabInfo.label}</span>;
+    }
+
+    tabs.push(tab);
+  });
```

Comment:
Here, it may be nicer to extract the status to a component. I'd add `useMemo` again and maybe make it declarative again (with `tabLabels.map`).

Generally, there is a lot of logic directly in the component render body, but that's normal when prototyping. Later, it's better use add `useMemo` and `useCallback` or even better — extract some of it to a custom hook.

Reply:
I refactored this using `useMemo` and a `map`, but I'm not sure I completely follow what it would entail to convert it to its own component or use a custom hook. Could you please elaborate a little bit about what that might look like so I can further improve this?

Reply:
For instance, the workflow grouping can be extracted simply as follows:

```ts
const useGroupedWorkflows = (
  workflows: Workflow[],
): Record<string, Workflow[]> => {
  return useMemo(() => {
    const grouped = {};
    workflows.forEach((workflow) => {
      if (!grouped[workflow.status]) {
        grouped[workflow.status] = [];
      }
      grouped[workflow.status].push(workflow);
    });
    return grouped;
  }, [workflows]);
};
```

This is a custom hook, it's as simple as that. Thanks to `useMemo`, it returns a stable identity for the same `workflows` parameter.

---

As for tabs and table headers — they don't have any dependencies from the state or component body, so they could be simple module-level arrays (indeed ever allocated only once):

```ts
const tabs = [
  {
    name: 'all',
    title: 'All',
    className: 'mailpoet-tab-all',
  },
  {
    name: WorkflowStatus.ACTIVE,
    title: 'Active',
    className: 'mailpoet-tab-active',
  },
  {
    name: WorkflowStatus.INACTIVE,
    title: 'Inactive',
    className: 'mailpoet-tab-inactive',
  },
  {
    name: WorkflowStatus.DRAFT,
    title: 'Draft',
    className: 'mailpoet-tab-draft',
  },
  {
    name: WorkflowStatus.TRASH,
    title: 'Trash',
    className: 'mailpoet-tab-trash',
  },
] as const;
```

I even simplified it by inlining the `className`, seems more redable to me now. (Another nice thing is we can add the `as const` and TS will now understand that `label: 'Trash'` belongs to `className: 'mailpoet-trash-tab'` instad of inferring `label: string` and `className` string. But we don't need this in the code below so not important.)

Similarly for the table headers...

Reply:
> This is a custom hook, it's as simple as that

Is the benefit here that we would be able to extract the callback outside of this component entirely? If so where would be the best place for it to live? Or is there a different advantage that I'm not seeing?

Reply:
Thank you again for your detailed explanation! I wasn't familiar with `as const` before.

Reply:
> So ultimately, it's just a tiny detail, nothing we'd have to do now

Still great to know, thanks for the explanation!